File: credentials-verification-android/app/src/main/java/io/iohk/cvp/viewmodel/MainViewModel.java
package io.iohk.cvp.viewmodel;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import javax.inject.Inject;

import io.iohk.cvp.grpc.AddConnectionFromTokenRunnable;
import io.iohk.cvp.grpc.AsyncTaskResult;
import io.iohk.cvp.grpc.GrpcTask;
import io.iohk.prism.protos.AddConnectionFromTokenResponse;
import io.iohk.prism.protos.ConnectorPublicKey;

public class MainViewModel extends CvpViewModel {

  private MutableLiveData<AsyncTaskResult<AddConnectionFromTokenResponse>> newConnectionInfo = new MutableLiveData<>();

  @Inject
  public MainViewModel() {
  }

  public LiveData<AsyncTaskResult<AddConnectionFromTokenResponse>> addConnectionFromToken(
          String token,
          ConnectorPublicKey publicKey, String nonce) {
    new GrpcTask<>(new AddConnectionFromTokenRunnable(newConnectionInfo), context)
        .execute(null, token, publicKey, nonce);
    return newConnectionInfo;
  }

}

File: credentials-verification-android/app/src/main/java/io/iohk/atala/prism/app/viewmodel/MainViewModel.kt
package io.iohk.atala.prism.app.viewmodel

import androidx.lifecycle.*
import io.iohk.atala.prism.app.data.DataManager
import io.iohk.atala.prism.app.data.local.db.mappers.ContactMapper
import io.iohk.atala.prism.app.data.local.db.model.ProofRequestWithCredentials
import io.iohk.atala.prism.app.grpc.AsyncTaskResult
import io.iohk.atala.prism.app.neo.common.EventWrapper
import io.iohk.atala.prism.app.utils.CryptoUtils
import io.iohk.atala.prism.protos.AddConnectionFromTokenResponse
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import javax.inject.Inject

class MainViewModel @Inject constructor(private val dataManager: DataManager) : NewConnectionsViewModel(dataManager) {

    companion object {
        private const val SYNC_TIME_IN_MILLISECONDS = 7000L
    }

    val requestSync: LiveData<EventWrapper<Boolean>> = liveData {
        while (true) {
            emit(EventWrapper(true))
            delay(SYNC_TIME_IN_MILLISECONDS)
        }
    }

    private val proofRequests: LiveData<List<ProofRequestWithCredentials>> = dataManager.allProofRequest()

    val proofRequest: LiveData<EventWrapper<ProofRequestWithCredentials>> = Transformations.map(proofRequests) {
        return@map if (it.isNotEmpty()) EventWrapper(it[0]) else null
    }

    private val _newConnectionInfoLiveData = MutableLiveData<AsyncTaskResult<AddConnectionFromTokenResponse>>()

    fun addConnectionFromToken(token: String, nonce: String)
            : LiveData<AsyncTaskResult<AddConnectionFromTokenResponse>>? {
        viewModelScope.launch(Dispatchers.IO) {
            try {
                val currentIndex = dataManager.getCurrentIndex()
                val addConnectionFromTokenResponse = dataManager.addConnection(dataManager.getKeyPairFromPath(CryptoUtils.getNextPathFromIndex(currentIndex)), token, nonce)
                dataManager.saveContact(ContactMapper.mapToContact(addConnectionFromTokenResponse.connection, CryptoUtils.getNextPathFromIndex(currentIndex)))
                dataManager.increaseIndex()
                _newConnectionInfoLiveData.postValue(AsyncTaskResult(addConnectionFromTokenResponse))
            } catch (ex: Exception) {
                _newConnectionInfoLiveData.postValue(AsyncTaskResult(ex))
            }
        }
        return _newConnectionInfoLiveData
    }
}